src/msg/simple/Pipe.cc

```diff
@@ -1336,7 +1336,11 @@ void Pipe::fault(bool onread)
     ldout(msgr->cct,10) << "fault already connecting, reader shutting down" << dendl;
     return;
   }
-  
+  if (errno == ECONNREFUSED) {
+    ldout(msgr->cct, 10) << "connection refused!" << dendl;
+    if (connection_state->clear_pipe(this))
+      msgr->dispatch_queue.queue_refused(connection_state.get());
+  }
```

Comment:
On Tue, Apr 12, 2016 at 07:35:43AM -0700, Gregory Farnum wrote:
> I haven't done a careful review, but the idea of this patch looks good.
> 
> This part isn't really safe though. We can get errors which don't directly call fault() or meander a bit before ending up here. We'll want some kind of shutdown error flag we can set at the failure site and check here or elsewhere.

I'll look into this, but I'm quite sure that ECONNREFUSED here would only
occur after attempt to connect() to port that isn't listening.